kura/org.eclipse.kura.linux.net/src/main/java/org/eclipse/kura/linux/net/dns/LinuxDns.java

```diff
@@ -171,149 +148,63 @@ public void setDnServers(Set<IPAddress> servers) {
-                                        servers.add(IPAddress.parseHostAddress(nameServer));
-                                    }
-                                    break;
-                                } else if (line.indexOf("expire") >= 0) {
-                                    // no DNS for this entry
-                                    break;
-                                }
-                            }
-                        }
-                    }
-                } catch (Exception e) {
-                    throw new KuraException(KuraErrorCode.INTERNAL_ERROR, e);
-                } finally {
-                    if (br != null) {
-                        try {
-                            br.close();
-                            br = null;
-                        } catch (IOException e) {
-                            e.printStackTrace();
+                servers = parseDhclientFile(globalDhClientFile, interfaceMatch, fixedAddressMatch);
+            } else {
+                servers = new ArrayList<>();
+                File dhclientFile = new File(formInterfaceDhclientLeasesFilename(interfaceName));
+                if (!dhclientFile.exists()) {
+                    try {
+                        if (dhclientFile.createNewFile()) {
+                            logger.debug("The {} doesn't exist, created new empty file ...", dhclientFile.getName());
                         }
+                    } catch (Exception e) {
+                        throw new KuraException(KuraErrorCode.CONFIGURATION_ERROR, e);
                     }
                 }
+            }
+        }
 
-            } else {
-                try {
-                    File dhclientFile = new File(formInterfaceDhclientLeasesFilename(interfaceName));
-                    if (!dhclientFile.exists()) {
-                        dhclientFile.createNewFile();
-                    }
-                    br = new BufferedReader(new FileReader(dhclientFile));
+        return servers;
+    }
 
-                    // boolean lookingForDns = false;
-                    String line;
+    private ArrayList<IPAddress> parseDhclientFile(File dhClientFile, String interfaceMatch, String fixedAddressMatch)
+            throws KuraException {
+        ArrayList<IPAddress> servers = new ArrayList<>();
+        try (FileReader fr = new FileReader(dhClientFile); BufferedReader br = new BufferedReader(fr)) {
+            String line = null;
+            while ((line = br.readLine()) != null) {
+                if ("lease {".equals(line.trim())) {
+                    // make sure it matches
+                    if (!br.readLine().trim().equals(interfaceMatch)) {
+                        continue;
+                    }
+                    if (!br.readLine().trim().equals(fixedAddressMatch)) {
+                        continue;
+                    }
                     while ((line = br.readLine()) != null) {
```

Comment:
can this, in some way, simplified/merged with 177 in order to have a single while that goes through the file?